solutions/advanced/cses-1110.mdx

```diff
@@ -0,0 +1,94 @@
+---
+id: cses-1110
+source: CSES
+title: Minimal Rotation
+author: Mihnea Brebenel
+---
+
+## Explanation
+
+We'll maintain the smallest lexicographically rotation after it's starting index, let it be $i$.
+Then, the comparison between this one and another rotation, starting at index $j$, is done by finding
+the first different character. This can be quickly done by binary searching the value $d$ such that
+$S_{i+d} \neq S_{j+d}$. Furthermore, hashing can be used to compare the prefixes of the rotations.
+
+The algorithm is also known as [Booth's Algorithm](https://en.wikipedia.org/wiki/Lexicographically_minimal_string_rotation).
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N \cdot \log{N})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+// BeginCodeSnip{Hashing Template (from the module)}
+class HashedString {
+  private:
+	// change M and B if you want
+	static const long long M = 1e9 + 9;
+	static const long long B = 9973;
+
+	// pow[i] contains B^i % M
+	static vector<long long> pow;
+
+	// p_hash[i] is the hash of the first i characters of the given string
+	vector<long long> p_hash;
+
+  public:
+	HashedString(const string &s) : p_hash(s.size() + 1) {
+		while (pow.size() <= s.size()) { pow.push_back((pow.back() * B) % M); }
+
+		p_hash[0] = 0;
+		for (int i = 0; i < s.size(); i++) {
+			p_hash[i + 1] = ((p_hash[i] * B) % M + s[i]) % M;
+		}
+	}
+
+	long long get_hash(int start, int end) {
+		long long raw_val =
+		    (p_hash[end + 1] - (p_hash[start] * pow[end - start + 1]));
+		return (raw_val % M + M) % M;
+	}
+};
+vector<long long> HashedString::pow = {1};
+// EndCodeSnip
+
+int main() {
+	string s;
+	cin >> s;
+
+	int sz = s.size();  // just a shorthand
+	int ans = 0;
+
+	// Double the string to make it circular and avoid working with modulo
+	s += s;
+	HashedString pref(s);
+	for (int i = 0; i < sz; i++) {
+		// Binary search for the first different character
+		int lo = 0;
+		int hi = sz - 1;
+		while (lo <= hi) {
```

Comment:
j curious but why not use the binary search that we use in the binary search module? not that it really matters b/c the solution idea is still the same

Reply:
tangentially related but idt standardizing binary search is feasible, it's used in too many scenarios

Reply:
fair enough, i just copy and paste the guide template bc I'm lazy lmao